python/python/lance/dataset.py

```diff
@@ -78,6 +85,44 @@
     from .progress import FragmentWriteProgress
 
 
+class MergeInsertBuilder(object):
```

Comment:
I'm going to keep poking at this but here is my current problem in case you know a solution off the top of your head.  It's actually related to this comment and the one below it.

I need a mutable reference to the `lance._Dataset` instance because this execute call changes its `ds` property

I can't save a `PyRefMut` in the `_MergeInsertBuilder` struct because it is not sync (which makes sense)

Unfortunately, if I make a subclass, I cannot override the `__init__` method (because it seems pyo3 does some weird magic with the `__new__` method that bypasses the normal inheritance).

So either `execute` needs to change to take in a `LanceDataset` (which is clunky for the user) or I need to find someway to get the `__init__` overload working correctly.

Reply:
I ended up solving this and was able to remove the boilerplate.